collectors/osidb_scheduler/collectors.py

```diff
@@ -0,0 +1,69 @@
+"""
+OSIDB Collectors
+"""
+from celery.utils.log import get_task_logger
+from django.contrib.contenttypes.models import ContentType
+from django.db.models import OuterRef, Q, Subquery
+from django.db.models.functions import Cast
+
+from collectors.framework.models import Collector
+from osidb.mixins import Alert
+
+logger = get_task_logger(__name__)
+
+
+class StaleAlertCollector(Collector):
+    """
+    Stale Alert Collector
+    """
+
+    def collect(self):
+        """
+        collector run handler
+
+        On every run, this collector will check if the alert is
+        still valid by comparing the creation time of the alert
+        with the validation time of the Model.
+
+        If the creation time of the alert is older than
+        the validation time of the Model,
+        the alert is considered stale and will be deleted.
+        """
+        content_types = ContentType.objects.filter(
+            id__in=Alert.objects.values_list("content_type", flat=True).distinct()
+        )
+
+        logger.info(
+            f"Searching for stale alerts in {content_types.count()} content types"
+        )
+
+        query = Q()
+
+        for content_type in content_types:
+            model_class = content_type.model_class()
+
+            subquery = Subquery(
+                model_class.objects.filter(
+                    pk=Cast(OuterRef("object_id"), output_field=model_class._meta.pk)
+                )
+                .order_by("last_validated_dt")
+                .values("last_validated_dt")[:1]
+            )
+
+            query |= Q(
+                content_type=content_type,
+                created_dt__lt=subquery,
+            )
+
+        filtered_queryset = Alert.objects.filter(query)
+
+        if filtered_queryset.count() == 0:
+            return "No Stale Alerts Found"
+
+        logger.info(f"Found {filtered_queryset.count()} stale alerts")
+
+        deleted_alerts_count = filtered_queryset.delete()[0]
```

Comment:
I am thinking about atomicity, do we need to somehow enforce it here? :thinking:

Reply:
Well I'm not sure about it, alerts are not modified once they go stale, so nothing should conflict

Reply:
This job should be the only piece of code performing write operations on the stale alerts so there should be no danger of any conflict. I do not see a reason for handling the atomicity. And as the `delete` is performed on the `queryset` I would expect that it is atomic anyway :smile: